Add k8s credential provider support

stmcginnis opened the pull request:
**Issue number:**

Closes #1702

**Description of changes:**

This adds the capability to use Kubernetes image credential providers to
retrieve credentials to use when pulling images for container creation.

Initially we will only support the ecr-credential-provider, but things
are set up so we may add more providers in future updates.

---

Yet to do:

- [X] Include `ecr-credential-provider` binary
- [X] Add new config settings
- [X] Perform full testing and verification
- [X] Documentation on using this feature (basic docs, more detailed docs TBD)

---

**Testing done:**

- [X] For `ecr-credential-provider` binary: Added RPM files, built and deployed instance, verified the binary was in the expected location on the filesystem.
- [X] Addition of new settings: Build, deployed, set values by using `apiclient set`. Verified settings saved and available using `apiclient get` and viewing config settings written to filesystem.
- [X] `any-enabled` template helper: Addition of unit tests to validate various input. Created small test app to execute templating code and verify expected output.
- [x] Application of settings:
  - [X] default cred provider config file written.
  - [X] Setting values recreates file and restarts kubelet with correct parameters.
  - [X] Sonobuoy quick tests pass
  - [X] Validate use of ECR plugin for registry validation
  - [X] Tested with Kubernetes 1.20, 1.21, 1.22, and 1.23
  - [X] Tested with separate `aws-config` and `aws-credentials` as well as combined using only `aws-config`

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

bcressey commented on packages/kubernetes-1.20/credential-provider-config.yaml:
```diff
@@ -0,0 +1,26 @@
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: CredentialProviderConfig
+{{! This isn't right as the condition is more than just whether `providers` is "truey" but whether any are enabled }}
```
If kubelet tolerates an empty `providers` list in the YAML then this is probably "right enough" if not perfect.

bcressey commented on packages/kubernetes-1.20/credential-provider-config.yaml:
```diff
@@ -0,0 +1,26 @@
+{{#if this.enabled}}
+  - name: {{this.name}}
+    matchImages:
+{{#each this.imagepatterns}}
+      - {{this}}
+{{/each}}
+    defaultCacheDuration: {{default "12h" this.cacheduration}}
+    apiVersion: credentialprovider.kubelet.k8s.io/v1beta1
```
I'd prefer the usual kebab-case style for the settings fields:
```suggestion
{{#each this.image-patterns}}
      - {{this}}
{{/each}}
    defaultCacheDuration: {{default "12h" this.cache-duration}}
    apiVersion: credentialprovider.kubelet.k8s.io/v1beta1
```

stmcginnis commented:
Note: right now this contains some of John's work from https://github.com/bottlerocket-os/bottlerocket/pull/2378 since that will need to merge before the addition of `ecr-credential-provider`. Hopefully looking at the individual commits can help isolate the relevant things if someone wants to review parts of this while it's being worked on.

bcressey commented on packages/kubernetes-1.20/credential-provider-config.yaml:
```diff
@@ -0,0 +1,25 @@
+{{#if this.name == "ecr-credential-provider"}}
+    env:
+      - name: PATH
+        value: /bin:/usr/bin:/usr/local/bin
```
This should already be the default `PATH`.

arnaldo2792 commented on sources/models/src/modeled_types/credentialproviders.rs:
```diff
@@ -0,0 +1,154 @@
+    inner: String,
+}
+
+lazy_static! {
+    // List of valid, supported credential provider plugins. Update this list
+    // if/when we ever add or remove credential provider plugin binaries.
+    pub(crate) static ref CREDENTIALPROVIDER_PLUGINS: Vec<&'static str> = vec!["ecr-credential-provider"];
+}
```
Doesn't this work for you?

```suggestion
pub(crate) const CREDENTIALPROVIDER_PLUGINS: [&str; 1] = ["ecr-credential-provider"];
```

stmcginnis replied:
Great, that works. Still learning. ;)

arnaldo2792 replied:
Nice! I'm learning too (I wasn't sure this would fit your use case)

stmcginnis replied:
One small change I did need to make. Instead of:

```
[&str; 1]
```

I needed:

```
[&'static str; 1]
```

arnaldo2792 commented on packages/kubernetes-1.20/kubelet-config:
```diff
@@ -111,6 +111,7 @@ runtimeRequestTimeout: 15m
 featureGates:
```
Nit: for the lack of a better place to put this comment

The commit's subject doesn't include a subsystem, I think for this commit `packages` should be used.

arnaldo2792 commented on packages/kubernetes-1.21/credential-provider-config.yaml:
```diff
@@ -0,0 +1,26 @@
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: CredentialProviderConfig
+{{! This isn't right as the condition is more than just whether `providers` is "truey" but whether any are enabled }}
+{{#if settings.kubernetes.credential-providers.providers}}
```
Would it make sense to have a handlebars helper that iterates through the list:

```handlebars
{{#if (any-enabled settings.kubernetes.credential-providers.providers)}}
```

stmcginnis replied:
I wasn't sure about adding a helper that would be specific to one use case like this. But I think we probably need to do that in order to properly format the yaml.

arnaldo2792 commented on sources/models/shared-defaults/kubernetes-services.toml:
```diff
@@ -36,6 +37,10 @@ template-path = "/usr/share/templates/kubernetes-ca-crt"
 path = "/etc/systemd/system/kubelet.service.d/exec-start.conf"
 template-path = "/usr/share/templates/kubelet-exec-start-conf"
 
+[credential-provider-config.yaml]
```
You are missing the `configuration-files` prefix, and the `.` will conflict with `toml`'s syntax, instead you should use `-` as in other configurations.

arnaldo2792 commented on packages/ecr-credential-provider/Cargo.lock:
```diff
@@ -0,0 +1,7 @@
+# This file is automatically @generated by Cargo.
```
This file isn't necessary.

stmcginnis replied:
Not sure how it was generated before, but I removed it and ran `cargo update` and it was not recreated. So I've removed it.

arnaldo2792 replied:
It seems like it was re-added in following forced pushes? Also, for the lack of a better place for this nit, there is a typo in this commit message:

```
from teh cloud-provider-aws project.
# Fixed:
from the cloud-provider-aws project.
```

And I would avoid "your" here:

```
It uses your AWS credentials or IAM

# Suggestion, an assumed IAM role is still an AWS credential
It uses AWS credentials
```

bcressey replied:
These `Cargo.lock` files can show up if you run `cargo make` in the package directory rather than in the root.

stmcginnis replied:
Removed file and cleaned up commit message.

arnaldo2792 commented on packages/ecr-credential-provider/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+sha512 = "d82ab5f9b05db8bcaf864faf87597d0f17cb9a4e52b0d192611725bf7e1c1512fe604b4a0aea0761058eca8ed6c1d0a96f6575d745b6082b9ad78c1290504990"
+bundle-modules = [ "go" ]
+
+[build-dependencies]
+# None
```
If this isn't required, then you could remove the field(?)

arnaldo2792 commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,45 @@
+%global gitrev b9c8d82db9edc3be45f82f8d5320d006c62aa50c
+%global shortrev %(c=%{gitrev}; echo ${c:0:7})
+
+%global _missing_build_ids_terminate_build 0
+%define debug_package %{nil}
```
We have been putting this flags on the top of the RPM spec

arnaldo2792 commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,45 @@
+License: Apache-2.0
+URL: https://github.com/kubernetes/cloud-provider-aws
+Source: bundled-v%{gover}.tar.gz
+Source1: v%{gover}.tar.gz
```
The actual source of the package should be `Source`

stmcginnis replied:
Done

arnaldo2792 commented on sources/models/src/modeled_types/credentialproviders.rs:
```diff
@@ -51,23 +51,14 @@ mod test_cacheduration_value {
-            "120m",
-            "2h15m"
-        ] {
+        for ok in &["12h", "120m", "2h15m"] {
```
This file was created in a different commit, and it was formatted here. You should have the formatted version in the commit that created the file.

stmcginnis replied:
Thanks, it has since been corrected. I will have a new chain of commits pushed eventually as I work my way through validating each step to build up to the whole solution.

arnaldo2792 commented on variants/aws-dev/Cargo.toml:
```diff
@@ -52,3 +52,4 @@ iputils = { path = "../../packages/iputils" }
 strace = { path = "../../packages/strace" }
 tcpdump = { path = "../../packages/tcpdump" }
 chrony = { path = "../../packages/chrony" }
+ecr-credential-provider = { path = "../../packages/ecr-credential-provider" }
```
I think it should be safe to declare `ecr-credential-provider` as a dependency of each kubernetes package, the reason being that now the kubelet has `KubeletCredentialProviders` and it actually will depend on the binaries provided by this package.

stmcginnis replied:
:+1: that's actually how I have it now in my local tree. So only change to the variants is with the dependencies pulled in from k8s.

arnaldo2792 commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,45 @@
+
+%build
+%set_cross_go_flags
+make ecr-credential-provider GIT_VERSION=%{gover}
```
I read the `Makefile` and it seems like `ecr-credential-provider` is statically linked since in the `Makefile` `CGO_ENABLED` is set to `0`. Another flag set in the `Makefile` is `-ldflags="$(LDFLAGS)"` and that doesn't match what is used in `hotdog` (the other statically linked go package included in Bottlerocket), which is just `-ldflags "-s"`. Even if we don't want to set that value for `-ldflags`, it will be best to use what other go packages use (`-ldflags "${GOLDFLAGS}"`). Thus, I think it will be better if we don't use the `Makefile` and instead we craft the compilation command to set the flags to the values we need.

stmcginnis replied:
I'm really hesitant to do that. The project may make changes to how the binary is compiled. if they do so, they will update the `Makefile` to adjust to any changes. So building with `make ecr-credential-provider` shouldn't change, even if how the build is actually performed is changed. So the `make` command is our "API contract" for how to build the binary.

This also makes sure the binary we create and package with Bottlerocket matches with what the officially release binaries from the project are. We don't want to get in a situation where `ecr-credentail-provider` in bottlerocket has a subtly different behavior than what someone would get from pulling down the binary from the upstream project.

arnaldo2792 replied:
I understand it is a little odd to not honor the `Makefile` provided by the upstream sources, but there are already packages in Bottlerocket where we change the upstream behavior to add additional flags to improve the security of the binaries, since we might have a different opinion on what a good set of default flags is for a given binary. Example packages where we don't use `Makefile`s provided by the upstream sources include [docker-cli], [docker-engine], [containerd], [nvidia-container-toolkit]. We even go a step further in the [libnvidia-container] spec, and we patch the `Makefile` to do what works best for Bottlerocket.

[docker-cli]: https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/docker-cli/docker-cli.spec
[docker-engine]: https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/docker-engine/docker-engine.spec
[containerd]: https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/containerd/containerd.spec
[nvidia-container-toolkit]: https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/nvidia-container-toolkit/nvidia-container-toolkit.spec
[libnvidia-container]: https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/libnvidia-container/libnvidia-container.spec

In the `Makefile` of `ecr-credential-provider`, they are just overriding `ldflags` with whatever `LDFLAGS` has, which in Bottlerocket's build system might be empty or set to the wrong flags, since the packages should be built with `"%{_cross_ldflags}"` so that cross compilation works as expected.

> We don't want to get in a situation where ecr-credentail-provider in bottlerocket has a subtly different behavior than what someone would get from pulling down the binary from the upstream project.

I understand this concern, and there are packages in Bottlerocket where we couldn't use the flags we wanted. One example of this is [nvidia-k8s-device-plugin], where we couldn't add the hardening flags since otherwise the libraries the binary depends on weren't loaded. But I caught that problem through testing, to make sure the package works as expected.

[nvidia-k8s-device-plugin]: https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/nvidia-k8s-device-plugin/nvidia-k8s-device-plugin.spec#L28

stmcginnis replied:
OK, that seems reasonable. I'll update this to follow what we're doing elsewhere then.

arnaldo2792 commented on packages/kubernetes-1.20/kubernetes-1.20.spec:
```diff
@@ -25,6 +25,7 @@ Source7: kubelet-bootstrap-kubeconfig
 Source8: kubernetes-tmpfiles.conf
 Source9: kubelet-sysctl.conf
 Source10: prepare-var-lib-kubelet.service
+Source11: credential-provider-config.yaml
```
None of the other templated files included in Bottlerocket use the extension of the language used to write them, instead a `-` is used and no extension is provided.

stmcginnis committed:
kubelet: enable credential provider support

This adds the capability to use Kubernetes image credential providers to
retrieve credentials to use when pulling images for container creation.

Initially we will only support the ecr-credential-provider, but things
are set up so we may add more providers in future updates.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis commented:
Marking as ready to review, but still performing testing/validation. Everything should be in place though, so pending any testing issues found this should be good to go.

bcressey commented on packages/ecr-credential-provider/ecr-credential-provider-tmpfiles.conf:
```diff
@@ -0,0 +1 @@
+C /var/lib/kubelet/plugins/ecr-credential-provider - - - -
```
We might need to unconditionally remove the old file too, unless tmpfiles.d overwrites this. Also, would be good to check perms to make sure they're `0755`.

stmcginnis replied:
Will add a remove call first. Perms appear to be good:

```
bash-5.1# ls -al /var/lib/kubelet/plugins/ecr-credential-provider
-rwxr-xr-x. 1 root root 15101208 Sep 27 16:28 /var/lib/kubelet/plugins/ecr-credential-provider
```

bcressey commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,56 @@
+%build
+%set_cross_go_flags
+
+# Set CGO_ENABLED=0 to statically link ecr-credential-provider, since it runs inside containers that
+# may not have the glibc version used to compile it
```
This looks like the `hotdog-hotpatch` comment - does `ecr-credential-provider` run in a container?

stmcginnis replied:
I'll update the comment, but `CGO_ENABLED=0` is how it is built upstream. I think we want to keep that to make it a static binary.

bcressey commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,56 @@
+%global _missing_build_ids_terminate_build 0
```
This might not be necessary with a more typical `go build` invocation, like other specs, where `-ldflags "${GOLDFLAGS}"` is set.

stmcginnis replied:
I was getting an error about the binary file not having a build ID. I would need to find the issue link again, but it appears to be a known issue with RPM packaging of Go binaries. This was the accepted workaround to prevent the error.

That said, there were a few things I was working through at the time, so I will try removing this and see if it is no longer an issue.

stmcginnis replied:
Just removing the top statement:

```
  #19 12.25 RPM build errors:
  #19 12.25     Missing build-id in /home/builder/rpmbuild/BUILDROOT/bottlerocket-x86_64-aws-ecr-credential-provider-1.24.1-1.aarch64/x86_64-bottlerocket-linux-gnu/sys-root/usr/share/factory/var/lib/kubelet/plugins/ecr-credential-provider
  #19 12.25     Generating build-id links failed
```

Adding `GOLDFLAGS`:

```
  #19 8.461 loadinternal: cannot find runtime/cgo
  #19 8.461 /usr/libexec/go/pkg/tool/linux_arm64/link: running x86_64-bottlerocket-linux-gnu-gcc failed: exit status 1
  #19 8.461 /usr/lib/gcc/x86_64-bottlerocket-linux-gnu/10/../../../../x86_64-bottlerocket-linux-gnu/bin/ld: cannot find Scrt1.o: No such file or directory
  #19 8.461 /usr/lib/gcc/x86_64-bottlerocket-linux-gnu/10/../../../../x86_64-bottlerocket-linux-gnu/bin/ld: cannot find crti.o: No such file or directory
  #19 8.461 collect2: error: ld returned 1 exit status
```

Removing `CGO_ENABLED=0`:

```
  #19 0.605 + go build -mod=vendor -trimpath -a -ldflags '-compressdwarf=false -linkmode=external' -o=ecr-credential-provider cmd/ecr-credential-provider/main.go cmd/ecr-credential-provider/main_test.go cmd/ecr-credential-provider/plugin.go cmd/ecr-credential-provider/plugin_test.go
  #19 2.553 # runtime/cgo
  #19 2.553 _cgo_export.c:3:10: fatal error: stdlib.h: No such file or directory
  #19 2.553     3 | #include <stdlib.h>
  #19 2.553       |          ^~~~~~~~~~
  #19 2.553 compilation terminated.
```

So looks like it may need to stay as it is. Unless there's another part I'm missing here.

bcressey replied:
Hmm, I will need to look at this - it indicates that something isn't being built with our expected toolchain. Those linker errors or missing headers shouldn't happen when the build flags are correct.

bcressey replied:
Hmm, I will need to look at this - it indicates that something isn't being built with our expected toolchain. Those linker errors or missing headers shouldn't happen when the build flags are correct.

bcressey replied:
Ah, of course - the spec needs a dependency on glibc for cgo and the normal linker flags to work.

```
diff --git a/packages/ecr-credential-provider/Cargo.toml b/packages/ecr-credential-provider/Cargo.toml
index 35e255f6..eb488b7f 100644
--- a/packages/ecr-credential-provider/Cargo.toml
+++ b/packages/ecr-credential-provider/Cargo.toml
@@ -15,3 +15,6 @@ releases-url = "https://github.com/kubernetes/cloud-provider-aws/releases"
 url = "https://github.com/kubernetes/cloud-provider-aws/archive/b9c8d82db9edc3be45f82f8d5320d006c62aa50c/v1.24.1.tar.gz"
 sha512 = "d82ab5f9b05db8bcaf864faf87597d0f17cb9a4e52b0d192611725bf7e1c1512fe604b4a0aea0761058eca8ed6c1d0a96f6575d745b6082b9ad78c1290504990"
 bundle-modules = [ "go" ]
+
+[build-dependencies]
+glibc = { path = "../glibc" }
diff --git a/packages/ecr-credential-provider/ecr-credential-provider.spec b/packages/ecr-credential-provider/ecr-credential-provider.spec
index 235aeb93..c7d5be91 100644
--- a/packages/ecr-credential-provider/ecr-credential-provider.spec
+++ b/packages/ecr-credential-provider/ecr-credential-provider.spec
@@ -1,6 +1,3 @@
-%global _missing_build_ids_terminate_build 0
-%define debug_package %{nil}
-
 %global goproject github.com/kubernetes
 %global gorepo cloud-provider-aws
 %global goimport %{goproject}/%{gorepo}
@@ -8,6 +5,8 @@
 %global gover 1.24.1
 %global rpmver %{gover}
 
+%global _dwz_low_mem_die_limit 0
+
 %global gitrev b9c8d82db9edc3be45f82f8d5320d006c62aa50c
 %global shortrev %(c=%{gitrev}; echo ${c:0:7})
 
@@ -23,6 +22,8 @@ Source1: bundled-v%{gover}.tar.gz
 Source100: root-.aws.mount
 Source1000: clarify.toml
 
+BuildRequires: %{_cross_os}glibc-devel
+
 %description
 %{summary}.
 
@@ -32,9 +33,7 @@ Source1000: clarify.toml
 
 %build
 %set_cross_go_flags
-
-# Set CGO_ENABLED=0 to statically link ecr-credential-provider.
-CGO_ENABLED=0 go build ${GOFLAGS} -trimpath -a -ldflags "-s" -o=ecr-credential-provider cmd/ecr-credential-provider/*.go
+go build -buildmode=pie -ldflags="${GOLDFLAGS}" -o=ecr-credential-provider cmd/ecr-credential-provider/*.go
 
 %install
 install -d %{buildroot}%{_cross_libexecdir}/kubernetes/kubelet/plugins
diff --git a/variants/Cargo.lock b/variants/Cargo.lock
index 1c64e2a6..b2993a37 100644
--- a/variants/Cargo.lock
+++ b/variants/Cargo.lock
@@ -285,6 +285,9 @@ dependencies = [
 [[package]]
 name = "ecr-credential-provider"
 version = "0.1.0"
+dependencies = [
+ "glibc",
+]
 
 [[package]]
 name = "ecs-agent"
```

You should also drop `-trimpath -a -ldflags "-s"` so that the debuginfo is added. (It will get removed later as part of the standard debuginfo stripping that happens in an rpm build.)

stmcginnis replied:
That works! Thanks for digging in to that.

bcressey commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,56 @@
+CGO_ENABLED=0 go build ${GOFLAGS} -trimpath -a -ldflags "-s" -o=ecr-credential-provider cmd/ecr-credential-provider/*.go
+
+%install
+rm -f %{_cross_sharedstatedir}/kubelet/plugins/ecr-credential-provider
```
Shouldn't need this line:
```suggestion
```

stmcginnis replied:
Removed!

bcressey commented on packages/kubernetes-1.20/credential-provider-config-yaml:
```diff
@@ -0,0 +1,27 @@
+      - name: HOME
+        value: /root
+      - name: AWS_SHARED_CREDENTIALS_FILE
+        value: /etc/kubernetes/kubelet/.aws/config
```
`/etc/aws/config` is probably a more standard way to render this. Since the file could have secrets, we would also want to ensure that the directory is mounted with the `secret_t` label, like `/etc/containerd` is.

stmcginnis replied:
Capturing here - Ben and I discussed this and decided to move the config to the more common location of the home directory. So latest version moves things to `/root/.aws/config`.

bcressey commented on packages/kubernetes-1.20/aws-config:
```diff
@@ -0,0 +1,3 @@
+{{~#if settings.kubernetes.credential-providers.aws-config~}}
+{{base64_decode settings.kubernetes.credential-providers.aws-config}}
+{{~/if~}}
```
It might be worth supporting both an `aws-config` and an `aws-credentials` file.

stmcginnis commented:
Just an update, this is currently blocked on this upstream issue:

https://github.com/kubernetes/kubernetes/issues/112842

I will track that and follow up here as that progresses.

bcressey commented on README.md:
````diff
@@ -441,6 +441,30 @@ The following settings are optional and allow you to further configure your clus
+  Example user data for setting up image credential providers:
+
+  ```toml
+  [settings.kubernetes.credential-providers]
+  # AWS config settings for the "default" profile
+  # Example command: "base64 ~/.aws/config"
+  aws-config = "base64_encoded_content"
+  # Example command: "base64 ~/.aws/credentials"
+  # (optional - credentials can be included in aws-config content)
+  aws-credentials = "base64_encoded_content"
````
My only real misgiving with this change is with these fields - it's going to be rather awkward if we end up needing AWS creds to be specified in some other context ("just put your creds for CloudWatch logging in the kubelet credential providers field").

What do you think about this as an alternative?
* Add these fields to `AwsSettings`
* Add `AwsSettings` to all models
* Conditionalize the `region` field in the struct with `#[cfg(variant_platform = "aws")]`

This would also let us simplify `[[settings.kubernetes.credential-providers.providers]]` down to just `[[settings.kubernetes.credential-providers]]` which would be more intuitive.

The main drawback I see is that we'd need to keep in mind whether a struct field in `AwsSettings` was "inside AWS" or "anywhere" in the future, but that seems OK.

stmcginnis replied:
That was my original take, but a certain tech lead asked for these settings to be a base64 encoded blob. ;)

I was/am concerned about any potential future functionality that would require AWS creds. If we wanted to keep with the blob route, I see a couple options:

- Change location to something specific for credential providers, then pass along environment variables to get it to read the credentials from somewhere other than `~/.aws`
- Allow multiple sub-settings to use AWS creds, but merge them into one file and require each be given a different profile name to use

The challenge with using explicit structs for the AWS creds is we will end up having three different options (at least so far) for how these AWS creds are configured:

- Left blank so instance role is used
- Populated with normal content a la `aws configure`
- Configured to use IAM Roles Anywhere with content like the following:
  ```
   [default]
   credential_process = ./aws_signing_helper credential-process \
      --certificate /path/to/certificate \
      --private-key /path/to/private-key \
      --trust-anchor-arn arn:aws:rolesanywhere:region:account:trust-anchor/TA_ID \
      --profile-arn arn:aws:rolesanywhere:region:account:profile/PROFILE_ID \
      --role-arn arn:aws:iam::account:role/role-name-with-path
  ```

We could expand our `settings.aws` to accomodate either of those (`settings.aws.config` and `settings.aws.credential-process`?). Maybe also require a profile name to be assigned to them, then require the credential provider to also provide the profile name to help differentiate which of the configured settings to use.

A few caveats here, so let me know if we should schedule a call to talk them through.

stmcginnis replied:
Adding what was discussed here to capture it for everyone.

We decided a good way to go would be to add an `enum` value that could either be a single `aws.profile` string, or a lookup of `aws.profile.[name]` values. So something like:

```toml
[settings.aws]
profile = default
```

or

```toml
[settings.aws.profile]
ecr-credential-provider = foo
future-service = default
```

I just pushed up changes to support that, though I'm running in to some issues. If I set the latter, then try to go back and set a single value, `thar-be-settings` complains that `profile` is a directory. It looks like it would need to have the smarts to delete this existing datastore location if someone wanted to switch back and forth between the two.

I'm also having problems pulling in `settings.aws.profile` into the `credential-provider-config.yaml` file. For some reason even though I see it set with `apiclient get settings.aws`, the evaluation always fails in the template so it never sets its `AWS_PROFILE` setting.

One more thing I noticed. With moving the `config` and `credentials` under `settings.aws`, for some reason it isn't detecting when those settings change and writing out the files under `/root/.aws`. I have those files still under kubernetes-settings, for lack of a more obvious correct place, so I would have expected changing those values to write out the rendered template and restart the kubelet service. It's only after changing `settings.kubernetes.credential-provider` settings that they get written out and the service gets restarted.

So still several things wrong with the current changes here. It feels like things are close, so I'm hoping I'm just missing some of the glue to pull these parts together.

stmcginnis replied:
I was having a hard time getting the switch between the single `profile` as a string and `profile` as a mapping of names to values. I think it might take some changes to schnauzer or thar-be-settings to allow it since it was an error that `profile` was a directory when it expected it to be a file. So for now I have set `profile` to only allow a single profile name to use for all services. We should be able to follow up with the combined string or key/value approach if and when we determine we actually need that.

With that change, I think this is very close. I still have the issue that any updates to `settings.aws` don't appear to trigger a restart of kubelet nor the writing out of the `aws-config` or `aws-credentials` templates under `/root/.aws`. I'm sure I'm just missing some wiring in the right place, but I may need some help figuring out where that would be.

webern replied:
> Conditionalize the region field in the struct with #[cfg(variant_platform = "aws")]

Hmmmm... not sure we should take our conditional compilation to that level of granularity.

stmcginnis replied:
I did try it out and it caused some compilation errors. I am not sure if there are any non-AWS local use cases where region might be needed to, so it may be safer to keep that for now.

stmcginnis commented:
Rebased on latest `develop` to pick up recent changes. One change was the addition of kubernetes 1.24. This now adds the plumbing to enable credential provider support for that as well.

Still `settings.aws` issues that need to be sorted.

arnaldo2792 commented on packages/ecr-credential-provider/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+releases-url = "https://github.com/kubernetes/cloud-provider-aws/releases"
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/kubernetes/cloud-provider-aws/archive/b9c8d82db9edc3be45f82f8d5320d006c62aa50c/v1.24.1.tar.gz"
```
They just released 1.25.1, should we jump to that version? Also, is there any documentation about the support for older k8s versions?

stmcginnis replied:
To be safe we could pull in the `cloud-provider-aws` release that corresponds to each k8s release. It doesn't look like there has been any change in the credential provider plugin interface, so for simplicity I was hoping we could avoid going down that route.

If folks think I'm being too optimistic with that approach, just let me know.

I will get this updated to pull in the new release.

arnaldo2792 commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,49 @@
+%build
+%set_cross_go_flags
+
+# Set CGO_ENABLED=0 to statically link ecr-credential-provider.
```
This comment isn't accurate anymore.

stmcginnis replied:
Removed

arnaldo2792 commented on packages/ecr-credential-provider/ecr-credential-provider.spec:
```diff
@@ -0,0 +1,49 @@
+%set_cross_go_flags
+
+# Set CGO_ENABLED=0 to statically link ecr-credential-provider.
+go build -buildmode=pie -ldflags="${GOLDFLAGS}" -o=ecr-credential-provider cmd/ecr-credential-provider/*.go
```
Do you need the `*.go` at the end? In the `hotdog` spec, we just specified the directory that contains the code.

stmcginnis replied:
I think that works for `hotdog` since there is only the `main.go` file in those directories. For this project there are multiple files that need to be compiled, and `go build` isn't happy if we just point to the parent directory.

arnaldo2792 commented on README.md:
````diff
@@ -951,10 +971,24 @@ They can be overridden for testing purposes in [the same way as other settings](
+  profile = "myprofile"
+  ```
+
+  3. Default settings. If `settings.aws.profile` is not set, or if there are named profile settings that do not include a matching name for the service you are configuring, they will fallback to using the "default" profile.
````
I didn't see any other numbered list for this section, is the `3.` extra here?

stmcginnis replied:
Oops, looks like some editor I used must have tried to be too "helpful" with editing this list. Fixed.

arnaldo2792 commented on README.md:
```diff
@@ -951,10 +971,24 @@ They can be overridden for testing purposes in [the same way as other settings](
 
-AWS-specific settings are automatically set based on calls to the Instance MetaData Service (IMDS).
+* `settings.aws.config`: The base64 encoded content to use for AWS configuration (e.g. `base64 ~/.aws/config`).
+* `settings.aws.credentials`: The base64 encoded content to use for AWS credentials (e.g. `base64 ~/.aws/credentials`).
```
I wouldn't use `base64 ~/.aws/<file>` as an example on how to obtain a value for this setting. Users may assume that they can set `credential_process` which allows arbitrary binaries to be used to fetch credentials, and in Bottlerocket such binaries aren't available (unless I'm wrong, and the ECR credential helper runs in a container with some tools).

I would like to understand a little bit more about the support of ephemeral credentials, are users going to be forced to update their user data every time the credentials stop working? Do we have a recommendation for them, i. e. use a host container that runs every X to refresh the credentials? Should we mention the limitations of using these APIs, like no support for `credential_process`?

stmcginnis replied:
We actually will be using `credential_process` later on. As a follow on, we have another issue for adding support for IAM Roles Anywhere. In order to use that functionality we will need to add a helper binary to Bottlerocket that they will then refer to in the config file.

This plugin is to get dynamic credentials, so they are expected to either use an instance profile, set hard coded credentials that don't expire, or use a helper binary like in the IAM Roles Anywhere case. We could theoretically add other helper binaries if needed. So we need the flexibility of allowing them to take their local configuration and just base64 encode it to set the content of these files.

arnaldo2792 commented on sources/models/src/modeled_types/credentialproviders.rs:
```diff
@@ -0,0 +1,132 @@
+use lazy_static::lazy_static;
```
Nit: shouldn't the name of this file be `credential_providers.rs`?

stmcginnis replied:
Renamed

arnaldo2792 commented on sources/models/src/modeled_types/credentialproviders.rs:
```diff
@@ -0,0 +1,132 @@
+string_impls_for!(CacheDurationValue, "CacheDurationValue");
+
+#[cfg(test)]
+mod test_cacheduration_value {
```
What's the reason why this isn't `test_cached_duration_value`? And similar comment in all other occurrences.

stmcginnis replied:
Yeah, changed.

arnaldo2792 commented on sources/models/src/modeled_types/credentialproviders.rs:
```diff
@@ -0,0 +1,132 @@
+lazy_static! {
+    // List of valid, supported credential provider plugins. Update this list
+    // if/when we ever add or remove credential provider plugin binaries.
+    pub(crate) static ref CREDENTIALPROVIDER_PLUGINS: [&'static str; 1] = ["ecr-credential-provider"];
```
hmm, didn't you said you had a way to avoid `lazy_static!`? 

https://github.com/bottlerocket-os/bottlerocket/pull/2377#discussion_r974814906

What changed?

stmcginnis replied:
Hmm, this concerns me. I did make that change, but it appears to have reverted back at some point. Changed back again, but will have to keep an eye out for anything else that changed.

arnaldo2792 commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -1,9 +1,9 @@
+use super::error;
```
Is this just a formatting change? Should you keep it out of this PR and instead fix any other formatting problem in a different PR?

stmcginnis replied:
Yeah, I think I had other changes that somehow triggered Clippy warnings, but have since been backed out. Plus I saw some imports were inserted in between the serde Error comment and the actual import. Will drop from here and do a quick little clean up PR.

arnaldo2792 commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -2362,3 +2397,71 @@ mod test_etc_hosts_helpers {
+    const TEMPLATE: &str = r#"{{#if (any-enabled input)}}enabled{{else}}disabled{{/if}}"#;
+
+    #[test]
+    fn any_enabled_enabled() {
```
Nit: I find these test names confusing, what about this: `test_any_enabled_with_enabled_elements`, `test_any_enabled_without_enabled_elements`, etc.

Also, it seems like all the tests have the same assertions.

stmcginnis replied:
Hmm, that's another concerning change that I can't explain.

Updated the names to be more descriptive and fixed tests to actually validate their intended cases.

arnaldo2792 commented on packages/kubernetes-1.20/kubelet-exec-start-conf:
```diff
@@ -17,6 +17,12 @@ ExecStart=/usr/bin/kubelet \
     --network-plugin cni \
     --root-dir /var/lib/kubelet \
     --cert-dir /var/lib/kubelet/pki \
+{{#if settings.kubernetes.credential-providers}}
+{{#if (any-enabled settings.kubernetes.credential-providers)}}
+    --image-credential-provider-bin-dir /usr/libexec/kubernetes/kubelet/plugins \
+    --image-credential-provider-config /etc/kubernetes/kubelet/credential-provider-config.yaml \
+{{/if}}
+{{/if}}
```
You could save an `if` statement if in the `any-enabled` helper evaluates `None` to falsey, what do you think?

stmcginnis replied:
Good call! This was originally needed when the structure was `settings.kubernetes.credential-providers.providers` since credential-providers would be null and cause an error. But now that we've simplified the structure I don't think there is any reason to have the extra check.

stmcginnis replied:
Turns out there actually was a reason I did it this way. If there is not a check before passing in the the handler there is an error rendering the file:

```
Error rendering "kubelet-exec-start-conf" line 20, col 1: `any_enabled` helper: Couldn't read parameter arg
```

I tried changing it to:

```
{{#if (and settings.kubernetes.credential-providers (any-enabled settings.kubernetes.credential-providers))}}
```

But it appears handlebars doesn't shortcircuit evaluation, so even though the first condition evaluates to falsey, it still tries to run the second part.

arnaldo2792 commented on packages/filesystem/aws-config:
```diff
@@ -0,0 +1,3 @@
+{{~#if settings.aws.config~}}
+{{base64_decode settings.aws.config}}
+{{~/if~}}
```
These two files (aws-config and aws-credentials) should be part of the `ecr-credential-provider` package, since they were added to support the functionality the package provides. 

If that doesn't "feel" right because there might be other packages that use said configurations I'll use the release package instead, since it contains files that are required by all images. `filesystem` should only be used to add new directories, and it shouldn't contain any file.

bcressey replied:
Either `release` or `os` will work as a package for files that should be in most or all variants.

stmcginnis replied:
Moved to `release` since we are planning to make these so they are there for other use cases besides `ecr-credential-provider`.

arnaldo2792 commented on packages/filesystem/root-.aws.mount:
```diff
@@ -0,0 +1,16 @@
+[Unit]
```
This mount should be part of the `release` package since it already contains mounts that are used to configure the system; and it should be closer to the configurations in [packages/release/etc-cni.mount](https://github.com/bottlerocket-os/bottlerocket/blob/1437bdade2fdf2b1a4cdde61fdcc2ae2624ff83c/packages/release/etc-cni.mount#L1)

arnaldo2792 commented on packages/kubernetes-1.24/Cargo.toml:
```diff
@@ -23,6 +23,7 @@ glibc = { path = "../glibc" }
 
 # RPM Requires
 [dependencies]
+ecr-credential-provider = { path = "../ecr-credential-provider" }
```
There is already a commit where the support for all the other variants was added. Please squash this commit with said commit.

arnaldo2792 commented on sources/models/shared-defaults/aws-creds.toml:
```diff
@@ -6,6 +6,7 @@ configuration-files = [
   "aws-config",
   "aws-credentials",
 ]
+restart-commands = []
```
This file is modified in two commits, please unify the changes.

arnaldo2792 commented on sources/models/shared-defaults/kubernetes-services.toml:
```diff
@@ -41,6 +43,14 @@ template-path = "/usr/share/templates/kubelet-exec-start-conf"
+[configuration-files.aws-credentials]
+path = "/root/.aws/credentials"
+template-path = "/usr/share/templates/aws-credentials"
+
```
This should be removed since it was added in `shared-defaults/aws-creds.toml`

bcressey commented on packages/kubernetes-1.23/credential-provider-config-yaml:
```diff
@@ -0,0 +1,25 @@
+    env:
+      - name: HOME
+        value: /root
+{{#if settings.aws.profile}}
+      - name: AWS_PROFILE
+        value: {{lookup-string-or-kv "default" "ecr-credential-provider" settings.aws.profile}}}}
+{{/if}}
```
Should the other `credential-provider-config-yaml` files use the same logic? Or is this a helper you were working on that you backed out?

stmcginnis replied:
I thought I had backed this all out, but apparently somehow missed this one. Will update.

bcressey commented on sources/api/migration/migrations/v1.11.0/aws-creds/src/main.rs:
```diff
@@ -0,0 +1,20 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+/// We added a new setting for AWS credential configuration.
+fn run() -> Result<()> {
+    migrate(AddSettingsMigration(&["settings.aws"]))
+}
```
This needs to be scoped to the individual settings since we previously had `settings.aws.region` for many variants.

I don't recall whether `settings.aws` will be fully cleaned up if all the settings underneath are removed. If not, and if that causes a problem, we might need to conditionalize this based on variant. `pki-affected-services` is the only other time we've done that in a migration.

stmcginnis replied:
> This needs to be scoped to the individual settings

Does this mean we need separate migrations for each property? So one each for `settings.aws.profile`, `settings.aws.config`, and `settings.aws.credentials`?

etungsten replied:
We have a `AddSettingsMigration` migration helper that accepts a list of settings: https://github.com/bottlerocket-os/bottlerocket/blob/aa3534b85c242919f148bd19f36d801790392b04/sources/api/migration/migration-helpers/src/common_migrations.rs#L6-L8

Example in use: https://github.com/bottlerocket-os/bottlerocket/blob/aa3534b85c242919f148bd19f36d801790392b04/sources/api/migration/migrations/archived/v1.0.8/kubelet-unsafe-sysctl-kube-reserved/src/main.rs#L9-L14

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -350,6 +351,12 @@ struct BootSettings {
 #[model]
 struct AwsSettings {
     region: SingleLineString,
```
Can you elaborate on the issue you ran into trying to make this a conditional field? I would really like it to be conditional on `#[cfg(variant_platform = "aws")]`.

You would need a custom build script like `api/early-boot-config/build.rs` has.

stmcginnis replied:
There was a pluto build error. That may just mean there would need to be some tweaks to how it is accessing that data. I have saved off that work in https://github.com/bottlerocket-os/bottlerocket/compare/develop...stmcginnis:bottlerocket:cred-providers-aws-profiles

But another concern is we may actually need this setting outside of running in an AWS variant. With IAM Roles Anywhere, and I would need to double check this, we may need to be able to point to a specific region for some configuration when running on meta- or vmware- variants.

webern replied:
> But another concern is we may actually need this setting outside of running in an AWS variant.

Something seems wrong about `settings.aws` for the credential provider. Did we explore any options to make the settings for the credential providers more 'generic'? Is it possible that `settings.aws.credentials` would be needed for something *other than* the k8s credential-provider plugin? Is there any possibility of deploying a credential provider with a host container or orchestrated container? I'm out of my depth here, but my hackles are up.

stmcginnis replied:
> Is it possible that settings.aws.credentials would be needed for something other than the k8s credential-provider plugin?

Yes, that's the key thing here. It's being added in support of the credential provider work, but we want to keep this somewhat separated so we can use this same mechanism if/when we add other features that work with AWS settings.

> Is there any possibility of deploying a credential provider with a host container or orchestrated container?

I don't believe there is any way to do that with the way the plugin mechanism works in kubelet.

webern replied:
Upon further reflection, I now understand why we arrived at `aws.settings.*` for this. It's because there's essentially one `~/.aws/` profile on the machine and it wouldn't make sense to bury that under, e.g. `settings.kubernetes`. Makes sense now: :+1:

bcressey replied:
> But another concern is we may actually need this setting outside of running in an AWS variant. With IAM Roles Anywhere, and I would need to double check this, we may need to be able to point to a specific region for some configuration when running on meta- or vmware- variants.

For use outside of EC2, I would expect the provided config to include a `region = ...` field as well. 

I could be swayed if we use this setting in the credential provider config, e.g.
```
{{#if settings.aws.region}}
      - name: AWS_REGION
        value: {{settings.aws.region}}
{{/if}}
```

Since that would remove one element of required configuration, and make it more likely for configs to do the right thing even inside EC2.

stmcginnis replied:
I had planned on doing this, but then thought of a corner case that might cause a problem by setting `AWS_REGION` in the credential provider config file.

Region is commonly set in the `~/.aws/config` file. When running outside of an AWS instance, that can be used by any clients to know which region to connect to. In those non-AWS instance cases, they may or may not set the `region` setting in user data. So for non-AWS deployments, setting `AWS_REGION` in the environment variables would give an option other than including it in `~/.aws/config`. It may cause some confusion as to which would be the one to take precedence,  but it could be an option.

The issue would be in any AWS instance deployments where the credentials defined in `~/.aws/config` defines and different region than where the instance is deployed. In this scenario, the `settings.aws.region` value will be set via SSM. So if we added that to the template to set `AWS_REGION` if `settings.aws.region` is set, that would *always* override whatever is in the config file. That would mean they would never be able to set the region to use to something other than where the instance is deployed.

I think we will have a use for setting `region` outside of AWS, but I don't think the current credential provider settings is the right place to use it.

May need to ponder this some more.

bcressey commented on sources/models/shared-defaults/kubernetes-services.toml:
```diff
@@ -36,6 +37,10 @@ template-path = "/usr/share/templates/kubernetes-ca-crt"
+[configuration-files.credential-provider-config-yaml]
+path = "/etc/kubernetes/kubelet/credential-provider-config.yaml"
+template-path = "/usr/share/templates/credential-provider-config-yaml"
+
```
This needs an `AddPrefixesMigration` to clean up the configuration-files entry - search in the archived migrations for related changes.

bcressey commented on sources/models/shared-defaults/aws-creds.toml:
```diff
@@ -0,0 +1,20 @@
+[metadata.settings.aws]
+affected-services = ["aws"]
+
+[configuration-files.aws-config]
+path = "/root/.aws/config"
+template-path = "/usr/share/templates/aws-config"
+
+[configuration-files.aws-credentials]
+path = "/root/.aws/credentials"
+template-path = "/usr/share/templates/aws-credentials"
```
Need an AddPrefixesMigration like this for the changes in this file:
```
fn run() -> Result<()> {
    migrate(AddPrefixesMigration(vec![
        "settings.aws.credentials",
        "settings.aws.config",
        "settings.aws.profile",
        "services.aws",
        "configuration-files.aws-config",
        "configuration-files.aws-credentials",
    ]))
}
```

Unfortunately, a rule of thumb is that almost any change reflected in these TOML model files needs to be covered by a migration.

(I think this can replace the `aws-creds` migration you have, but the `aws-creds-metadata` one is still needed to cover `metadata.settings.aws`.)

bcressey commented on sources/models/shared-defaults/kubernetes-services.toml:
```diff
@@ -7,6 +7,7 @@ configuration-files = [
   "kubelet-exec-start-conf",
   "kubernetes-ca-crt",
   "proxy-env",
+  "credential-provider-config-yaml",
```
This needs a `ReplaceListsMigration` to fix up the existing list on upgrade, and restore it on downgrade.

bcressey commented on sources/api/migration/migrations/v1.11.0/credential-providers/src/main.rs:
```diff
@@ -0,0 +1,24 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+/// We added a new setting for configuring kubelet's image credential
+/// provider plugins. Initially this is only to support ecr-credential-provider,
+/// but others may be added as needed.
+fn run() -> Result<()> {
+    migrate(AddSettingsMigration(&[
+        "settings.kubernetes.credential-providers",
+    ]))
+}
```
This should be an `AddPrefixesMigration` since there will be a cluster of settings underneath.

bcressey commented on sources/models/src/modeled_types/credentialproviders.rs:
```diff
@@ -0,0 +1,132 @@
+    inner: String,
+}
+
+lazy_static! {
+    // List of valid, supported credential provider plugins. Update this list
+    // if/when we ever add or remove credential provider plugin binaries.
+    pub(crate) static ref CREDENTIALPROVIDER_PLUGINS: [&'static str; 1] = ["ecr-credential-provider"];
```
Instead of a hard-coded list, what do you think about validating at runtime that the provided plugin name exists in the right place on the filesystem?

That would eliminate the need to touch this part of the code after packaging a plugin and including it in the image.

stmcginnis replied:
We could just remove this then. Kubelet will check that the named binary exists in the `--image-credential-provider-bin-dir` directory. Would that be preferable?

stmcginnis replied:
I've changed this input type to be similar to `SingleLineString` with the addition of checking for any path traversal characters. So we can at least make sure someone isn't trying to access something they shouldn't be, but we can leave the actual validation of the plugin binary to kubelet.

With the previous way, if the name was not correct it would actual stall the boot of an instance and it wasn't very apparent to the user what was happening and what they needed to do to resolve it. With simplifying our checks here, the node will now boot with an invalid plugin name, but kubelet will error and log a message that it can't find the plugin. Then the instance will at least be running and accessible, and the user has a better chance to view the logs and see what the root of the issue is.

stmcginnis commented:
I believe I have addressed all comments so far. Please let me know if I have missed anything.

I've resolved all of the ones that I think were fully addressed. There are a few that either had questions or were open to further discussion which I have not resolved yet pending some sort of resolution.

stmcginnis commented:
Something got messed up with all the rebasing to fix up commits with the various updates. AWS settings are now not being recognized. Investigating and will update soon.

stmcginnis commented:
I believe I've sorted out all the issues and I've been able to test a scenario on a 1.22 k8s cluster. All rendering is working as expected and plugin configuration works and is able to pull an image from a repo owned by another account.

matthewdaltonditto commented:
@arnaldo2792 pinging for re-review

stmcginnis commented:
Updated to resolve merge conflict in `sources/Cargo.lock`.

webern commented on README.md:
```diff
@@ -951,10 +971,24 @@ They can be overridden for testing purposes in [the same way as other settings](
 ##### AWS-specific settings
 
-AWS-specific settings are automatically set based on calls to the Instance MetaData Service (IMDS).
+* `settings.aws.config`: The base64 encoded content to use for AWS configuration (e.g. `base64 ~/.aws/config`).
```
Should consider mentioning that these settings are not necessary when relying on the instance role (either here or below where you have a `**Note**`)?

stmcginnis replied:
Good call, added a note.

webern commented on sources/api/schnauzer/src/lib.rs:
```diff
@@ -137,6 +137,7 @@ pub fn build_template_registry() -> Result<handlebars::Handlebars<'static>> {
     );
     template_registry.register_helper("localhost_aliases", Box::new(helpers::localhost_aliases));
     template_registry.register_helper("etc_hosts_entries", Box::new(helpers::etc_hosts_entries));
+    template_registry.register_helper("any-enabled", Box::new(helpers::any_enabled));
```
Oh man sorry, this is a mess. Above we added this comment

```
// Prefer snake case for helper names (we accidentally created a few with kabob case)
```

Should we follow it and change this to snake_case?

```suggestion
    template_registry.register_helper("any_enabled", Box::new(helpers::any_enabled));
```

stmcginnis replied:
I completely missed that comment and had thought I was keeping things more consistent by going with kabob case. Updating to snake case.

webern commented on sources/models/src/modeled_types/credential_providers.rs:
```diff
@@ -0,0 +1,128 @@
+    // A duration string is a sequence of decimal numbers, each with a
+    // unit suffix, such as "300ms" or "2h45m".
+    // Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+    // Based on restraints from Go's `ParseDuration` funciton.
```
```suggestion
    // Based on contraints from Go's `ParseDuration` function.
```

webern commented on sources/models/src/modeled_types/credential_providers.rs:
```diff
@@ -0,0 +1,128 @@
+    // unit suffix, such as "300ms" or "2h45m".
+    // Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+    // Based on restraints from Go's `ParseDuration` funciton.
+    pub(crate) static ref DURATION_FORMAT: Regex = Regex::new(r"^([0-9]*[a-z]+)+$").unwrap();
```
If we're constraining this string, does it makes sense to require the correct suffix? If I read this right, this is valid: `000xyz`.

stmcginnis replied:
Probably, but now based on the plugin name validation I'm wondering if we should be doing any of this validation at all, or if we should just pass along whatever string to the kubelet settings to be validated there.

stmcginnis replied:
Dropping this in favor of the existing `KubernetesDurationValue` which appears to be the same thing.

webern commented on sources/models/src/modeled_types/credential_providers.rs:
```diff
@@ -0,0 +1,128 @@
+
+    #[test]
+    fn invalid_cache_duration_value() {
+        for err in &["12 days", "💝", "Ï", ""] {
```
If tightening the validation based on my comment on the regex, then we would want to add something like:

```suggestion
        for err in &["12 days", "💝", "Ï", "", "1minute"] {
```

webern commented on sources/models/shared-defaults/aws-creds.toml:
```diff
@@ -0,0 +1,20 @@
+restart-commands = []
+
+[metadata.settings.aws]
+affected-services = ["aws"]
```
I thought the purpose of this was to restart a systemd service, e.g. in this case it would be `aws.service`, when the settings (and thus these files) change. In this case there is no service, right? Is this supposed to be an empty array?

Is this necessary, and if so, should is this the right name (and not e.g. `aws-profile`)?

stmcginnis replied:
This needs to be an empty array so things are updated but nothing is executed to restart a service.

webern commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1301,6 +1304,37 @@ pub fn etc_hosts_entries(
+        enabled: bool,
+    }
+
+    let mut result = false;
+    match arg {
+        Value::Array(items) => {
+            for item in &items {
+                let res = serde_json::from_value::<EnablableObject>(item.clone());
+                if let Ok(eo) = res {
+                    if eo.enabled {
+                        result = true;
+                        break;
+                    }
+                }
+            }
+        }
+        _ => ()
+    }
+    result
```
Not a very important nit, take it or leave it:

```suggestion
    match arg {
        Value::Array(items) => {
            for item in &items {
                let res = serde_json::from_value::<EnablableObject>(item.clone());
                if let Ok(eo) = res {
                    if eo.enabled {
                        return true;
                    }
                }
            }
        }
        _ => ()
    }
    false
```

stmcginnis replied:
If I need to touch this part again I will update, but if you don't feel strongly I'd like to avoid some rebasing and churn as I try to get everything addressed.

stmcginnis replied:
Now I remember why I went this way. I tried this out in the new PR for the helper and had some issues. This may be due to my lack of deep understanding of how macros work in Rust, but with the `handlebars_helper` macro we are not able to just `return true;`. That causes a compile time error with it expecting the returned type to be a `Result()`.

I'm not sure why we are able to return a boolean with the last statement just being `result` but can't explicitly return a bool. But the helper macro takes that and creates the correct return value.

So to directly return from within these loops, we would need to import a couple more structs and change the lines to:

```rust
use handlebars::{JsonValue, ScopedJson};
...
return Ok(ScopedJson::Derived(JsonValue::from(true)))
```

It seems more readable to me to have the current result variable, but if you would rather see it done the other way (or if you can educate me on rust with a better way to accomplish this!) please add that to the new PR for this part: https://github.com/bottlerocket-os/bottlerocket/pull/2541

webern commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -2362,3 +2397,74 @@ mod test_etc_hosts_helpers {
+
+    #[test]
+    fn test_any_enabled_with_enabled_elements() {
+        let result = setup_and_render_template(TEMPLATE, &json!([{"enabled": true}])).unwrap();
```
Seems like it would be good to prove that we don't care what else is in the object.

```suggestion
        let result = setup_and_render_template(TEMPLATE, &json!([{"foo": [], "enabled": true}])).unwrap();
```

webern commented on sources/api/migration/migrations/v1.11.0/aws-creds-metadata/src/main.rs:
```diff
@@ -0,0 +1,23 @@
+fn run() -> Result<()> {
+    migrate(AddMetadataMigration(&[SettingMetadata {
+        metadata: &["affected-services"],
+        setting: "settings.aws",
```
Here's that mysterious service again.

```
        setting: "settings.aws",
```

stmcginnis replied:
I forget the details now, but I was told this was needed to get the plumbing of everything working right.

webern reviewed (approved):
I guess I'm a little confused about what the "service" named `aws` is, but assuming that is done correctly... looks good.

bcressey commented on README.md:
````diff
@@ -421,6 +421,26 @@ The following settings are optional and allow you to further configure your clus
+
+  Example user data for setting up image credential providers:
+
+  ```toml
+  [[settings.kubernetes.credential-providers]]
+  enabled = true
+  name = "ecr-credential-provider"
+  # (optional - defaults to "12h")
+  cache-duration = "30m"
+  image-patterns = [
+    # One or more URL paths to match an image prefix. Supports globbing of subdomains.
+    "*.dkr.ecr.us-east-2.amazonaws.com",
+    "*.dkr.ecr.us-west-2.amazonaws.com"
+  ]
+  ```
````
Elsewhere we've preferred maps rather than lists to represent these "named" structures, e.g. `settings.host-containers` and `settings.pki`.

One of the main reasons is that it's more user-friendly to do targeted edits to a specific entry:
```
apiclient set kubernetes.credential-providers.ecr-credential-provider.enabled=true
```

With a list, I believe the entire list would need to be provided again, in the JSON syntax.

stmcginnis replied:
Done in https://github.com/bottlerocket-os/bottlerocket/pull/2547

bcressey commented on packages/ecr-credential-provider/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+releases-url = "https://github.com/kubernetes/cloud-provider-aws/releases"
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/kubernetes/cloud-provider-aws/archive/refs/tags/v1.25.1.tar.gz"
```
nit: I tend to prefer this form of the URL for consistency with other packages:
```suggestion
url = "https://github.com/kubernetes/cloud-provider-aws/archive/v1.25.1/cloud-provider-aws-1.25.1.tar.gz"
```

arnaldo2792 replied:
I wouldn't call it a nit, in the cache this will be named `1.25.1.tar.gz` which could collide with other files (if there are any).

etungsten replied:
To clarify, although not ideal, naming collisions in Bottlerocket's third-party lookaside cache is not an issue since the files are prefixed with their hash in the bucket and buildsys knows how to retrieve the file based on its hash. 

We already have naming collisions between all the EKS-D source tarballs for the different K8s packages since they're all named `kubernetes-src.tar`

bcressey replied:
I was thinking - and I think @arnaldo2792 was thinking - of hypothetical collisions within the spec file, where prefixing the version with the source project's name makes it clear which "v1.25.1.tar.gz" is being unpacked.

stmcginnis replied:
Will change.

Out of curiosity, how did you get that path? If you go to the GitHub repo for the project and look at the release artifacts, the link that it provides is the one I had used. Is there somewhere else that publishes the more descriptively named link? Or did you just know GitHub also provides a link that way?

bcressey commented on packages/filesystem/filesystem.spec:
```diff
@@ -55,6 +56,7 @@ ln -s lib %{buildroot}%{_cross_prefix}/lib64
 /dev
 /proc
 /root
+/root/.aws
```
nit: this is technically included by `/root` above, the same way `/media` includes `/media/cdrom` 
```suggestion
```

stmcginnis replied:
Without this line there is an error writing to the subdirectory that it is read only. I believe the subdirectory needs to be explicitly created in order for the mount operation to succeed.

bcressey replied:
> Without this line there is an error writing to the subdirectory that it is read only.

An error during the rpm build or at runtime? There shouldn't be a need to write to `/root` as long as the directory exists and is included in the package. Same as with `/media/cdrom` which is only used as a mount point at runtime, and only included implicitly because of the `/media` line.

stmcginnis replied:
During runtime. When the tmpfs is being mounted to `/root/.aws` it fails because the folder does not exist and cannot be created.

stmcginnis replied:
Error on boot:

```
[FAILED] Failed to mount AWS configuration directory (/root/.aws).
...
thar-be-settings[1497]: Failed to write template directory to disk at /root/.aws: Read-only file system (os error 30)
```

stmcginnis replied:
I was able to clean this up a bit in https://github.com/bottlerocket-os/bottlerocket/pull/2542 - PTAL.

arnaldo2792 commented on variants/Cargo.lock:
```diff
@@ -464,7 +464,6 @@ dependencies = [
 name = "kubernetes-1_24"
 version = "0.1.0"
 dependencies = [
- "ecr-credential-provider",
```
This commit removes `ecr-credential-provider` from `kubernetes-1.24` which was added in the first commit of the series, and then it is added back in a later commit. Please clean up so that it is added in one commit.

stmcginnis replied:
Is our "unit of change" consider a PR or a commit? I tried to put logical changes into separate commits to make it easier to review, but it's been a nightmare with needing to rebase and move things around to address comments. That has resulted in cases like this where git decides to remove and readd things.

I am going to break out this PR into multiple PRs to hopefully simplify reviewing and updating. Then we can start to get some pieces of this merged and out of the way and can focus on specific aspects of the overall change.

arnaldo2792 commented on sources/Cargo.lock:
```diff
@@ -2014,20 +2014,6 @@ dependencies = [
  "migration-helpers",
 ]
 
-[[package]]
```
These were added in the second commit of the series, and then removed here. Please clean up so that they only appear in the commit where they were added.

bcressey commented on sources/api/migration/migrations/v1.11.0/kubelet-aws-config/src/main.rs:
```diff
@@ -0,0 +1,29 @@
+    migrate(AddPrefixesMigration(vec![
+        "settings.aws.credentials",
+        "settings.aws.config",
+        "settings.aws.profile",
```
These first three are covered by the single-setting migrations in `aws-creds`, and don't need to be handled here.

bcressey commented on sources/api/migration/migrations/v1.11.0/kubelet-credential-providers/src/main.rs:
```diff
@@ -0,0 +1,23 @@
+    migrate(AddPrefixesMigration(vec![
+        "configuration-files.credential-provider-config-yaml",
+    ]))
+}
```
nit: this can get lumped in with the other `AddPrefixesMigration` in `credential-providers`, which is preferable since then there's only one binary to build and download.

bcressey commented on sources/api/migration/migrations/v1.11.0/aws-creds/src/main.rs:
```diff
@@ -0,0 +1,24 @@
+        "settings.aws.credentials",
+        "settings.aws.profile",
+    ]))
+}
```
In addition to these three settings, all non-aws variants will also pick up a potential new setting - `settings.aws.region` - that needs to be handled by a migration.

So your choices are:
* conditionalize the removal of that setting using `if cfg!(variant_platform = "aws")` (similar to what the `pki-affected-services` did)
* or conditionalize the existence of that setting in the non-aws models, so that it won't exist and doesn't need to be migrated

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -1128,3 +1128,59 @@ mod test_valid_kmod_key {
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// SingleFileName is the name of a file, such as a plugin or other binary on
```
nit: perhaps `SafeFileName` since it's more about the contents of the name than the quantity of names

stmcginnis replied:
Changing the `CredentialProvider` collection to be a hashmap, with the name of the plugin as the key name. So this is no longer needed.

bcressey commented on sources/models/src/modeled_types/credential_providers.rs:
```diff
@@ -0,0 +1,79 @@
+lazy_static! {
+    // A duration string is a sequence of decimal numbers, each with a
+    // unit suffix, such as "300ms" or "2h45m".
+    // Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+    // Based on constraints from Go's `ParseDuration` function.
+    pub(crate) static ref DURATION_FORMAT: Regex = Regex::new(r"^([0-9]*[a-z]+)+$").unwrap();
```
Is this the same as a `KubernetesDurationValue`?

stmcginnis replied:
Huh, it would appear so!

stmcginnis commented:
Rebased and resolved merge conflicts.

stmcginnis commented:
Thanks everyone!
